### PR TITLE
Add tags to groups

### LINKED ISSUE #229: Add Group Tags
For https://github.com/flywheel-io/frontend/issues/127

We need the ability to save tags to a group. If a tag is deleted it needs to be removed from that group's children. 

@rentzso and/or @nagem, anything else this needs?

### PR DESCRIPTION
**Pull request by nagem**
Closes #229

Changes:
 - Groups can be tagged
 - Deleting a group tag deletes the tag down the heirarchy
 - Renaming a group tag renames tags down the heirarchy
 - All tags must be 1-32 characters long

### CODE REVIEW

**Comment by mrDarcyMurphy** [comment]
@nagem I've only tested adding/editing/deleting on the group and I've found the following bug when deleting. It's deleting the tag, but it's responding with an error.

```
uwsgi_1          | 2016-04-11 20:39:13      scitran.api                  base.py   109:DEBU looked up cached token in 0ms
uwsgi_1          | 2016-04-11 20:39:13      scitran.api                  base.py   249:DEBU from None darcymurphy@flywheel.io DELETE /api/groups/scitran/tags/TESTING {u'root': u'true'}
uwsgi_1          | 2016-04-11 20:39:13      scitran.api            validators.py   135:DEBU {u'additionalProperties': False, u'$schema': u'http://json-schema.org/draft-04/schema#', u'type': u'object', u'properties': {u'value': {u'type': u'string'}}}
uwsgi_1          | 2016-04-11 20:39:13      scitran.api           liststorage.py    99:DEBU query_params TESTING
uwsgi_1          | 2016-04-11 20:39:13      scitran.api           liststorage.py   102:DEBU query {'_id': 'scitran'}
uwsgi_1          | 2016-04-11 20:39:13      scitran.api           liststorage.py   103:DEBU update {'$pull': {'tags': 'TESTING'}}
uwsgi_1          | 2016-04-11 20:39:13      scitran.api                  base.py   230:ERRO Traceback (most recent call last):
uwsgi_1          |   File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 570, in dispatch
uwsgi_1          |     return method(*args, **kwargs)
uwsgi_1          |   File "./api/handlers/listhandler.py", line 308, in delete
uwsgi_1          |     payload = self.request.json_body
uwsgi_1          |   File "/usr/local/lib/python2.7/dist-packages/webob/request.py", line 725, in _json_body__get
uwsgi_1          |     return json.loads(self.body.decode(self.charset))
uwsgi_1          |   File "/usr/lib/python2.7/json/__init__.py", line 338, in loads
uwsgi_1          |     return _default_decoder.decode(s)
uwsgi_1          |   File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
uwsgi_1          |     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
uwsgi_1          |   File "/usr/lib/python2.7/json/decoder.py", line 384, in raw_decode
uwsgi_1          |     raise ValueError("No JSON object could be decoded")
uwsgi_1          | ValueError: No JSON object could be decoded
uwsgi_1          |
uwsgi_1          | [pid: 16|app: 0|req: 59/126] 192.168.99.1 () {52 vars in 980 bytes} [Mon Apr 11 20:39:13 2016] DELETE /api/groups/scitran/tags/TESTING?root=true => generated 66 bytes in 30 msecs (HTTP/1.1 500) 3 headers in 130 bytes (2 switches on core 1)
```

File: api/handlers/listhandler.py [hunk 1]
```diff
@@ -275,6 +283,56 @@ def put(self, cont_name, list_name, **kwargs):
...
+            update = {'$set': {'tags.$': new_value}}
+            self._propagate_group_tags(_id, query, update)
+        return result
+
+    def delete(self, cont_name, list_name, **kwargs):
+        _id = kwargs.get('cid')
+        result = super(TagsListHandler, self).delete(cont_name, list_name, **kwargs)
+        if cont_name == 'groups':
+            payload = self.request.json_body
+            deleted_tag = payload.get('value')
```
Comment (nagem) [inline api/handlers/listhandler.py]: This should be replaced with `kwargs.get('value')`

**Comment by nagem** [comment]
@mrDarcyMurphy I commented where the issue is, will fix soon.

**Comment by mrDarcyMurphy** [comment]
Thanks. Everything else seems to be fine.

**Comment by nagem** [comment]
@mrDarcyMurphy Bug fixed, thanks for pointing it out!

**Comment by mrDarcyMurphy** [comment]
@nagem You're welcome. The fix worked, everything checks out on my end. 

If you want, check out the front end branch on [this PR](https://github.com/flywheel-io/frontend/pull/284) to try it all out. You'll be able to add tags to groups (under the site admin) and add them to sessions. Edits to tags show up in both places.

**Comment by nagem** [comment]
@rentzso let me know if you see anything that should be changed otherwise I'll merge in a bit as @mrDarcyMurphy is using this branch without any issues.

File: api/handlers/listhandler.py [hunk 2]
```diff
@@ -275,6 +283,55 @@ def put(self, cont_name, list_name, **kwargs):
...
+    def delete(self, cont_name, list_name, **kwargs):
+        _id = kwargs.get('cid')
+        result = super(TagsListHandler, self).delete(cont_name, list_name, **kwargs)
+        if cont_name == 'groups':
+            deleted_tag = kwargs.get('value')
+            query = {}
+            update = {'$pull': {'tags': deleted_tag}}
+            self._propagate_group_tags(_id, query, update)
+
+    def _propagate_group_tags(self, _id, query, update):
```
Comment (rentzso) [inline api/handlers/listhandler.py]: db operations should go in the `dao` module or in a submodule. We made some exceptions in the past, but it seems the right thing to do here (maybe together with `_propagate_project_permissions` in the permission handler).

**Comment by rentzso** [comment]
LGTM
Just a comment but feel free to merge this PR and open an issue for it.

File: api/handlers/listhandler.py (hunk 2 above)
Comment (nagem) [inline api/handlers/listhandler.py]: That is a very good point, I'll add it as a requirement of https://github.com/scitran/core/issues/81.

**Pull request merged**